Add global hooks to `torch.nn.Module`

emcastillo opened the pull request:
This allows registering hooks that will be executed for every module.

This idea arose in a discussion with @tkerola and @niboshi kindly proposed this approach.

The use case for this is to avoid boilerplate code when registering the same hook for all the modules in a complex model, the internal use-case was to allow every model to accept a NumPy array in the forward pass in a simpler way. Other use cases involve general mechanisms for plotting or tracing & debugging.

Currently, this is shared for all the modules but this can be worked out to have the hooks shared only per type of module.

If this functionality is not needed feel free to close the PR.

Commit:
Add global hooks to `torch.nn.Module`

albanD commented:
Hi,

Thanks for the PR.
I think I would have similar questions as the ones asked on https://github.com/pytorch/pytorch/pull/38604:
- How would this interact with the jit and ScriptModule?
- How would this interact with save/load?
- How would this interact with multiprocessing (including Dataloader)?
- Can you give a clear example of how you use it in your code? Is the global state really justified here?

emcastillo commented:
Hi!

> How would this interact with the jit and ScriptModule?
I have observed that regular hooks are disabled when a module is jitted. Currently the global hooks are executed when a jitted module forward is called but they are never called for internal modules. I will blacklist the global hooks to have consistent behavior.

> How would this interact with save/load?
Hooks are not being saved when a module is serialized and restored. The global hooks should follow the same policy.

> How would this interact with multiprocessing (including Dataloader)?
If the process is forked after registering the hook, It will be available in the child process. Otherwise, it won't be. I think that this is the same for all the other hooks.

Can you give a clear example of how you use it in your code? Is the global state really justified here?

Some use cases can be to convert between types of the module forward inputs.
```python
def convert_to_tensor(module, input):
    torch_inputs = [x for x in input]
    for i, x in enumerate(input):
        if not isinstance(x, torch.Tensor):
            torch_inputs[i] = torch.tensor(x, requires_grad=True)
    return tuple(torch_inputs)


torch.nn.Module.register_global_forward_pre_hook(convert_to_tensor)
module = MyModule()
new_module = MyModule2()
input = numpy.ones((5,), dtype=numpy.float32)
module(input)
module2(input)
```

Other use cases can be used for profiling or debugging, such as finding the module that created a nan in the output. (Thanks to @take-cheeze!).
This is mostly to reduce code for registering the same handle to several modules and do it at once.

However, I also feel that some global state can be a bit burdensome, so maybe we can use a context manager that is accessible using thread-local storage to easily enable and disable them? 
Something in the lines of

```python
global_hooks = torch.nn.module.Global_Hooks()
global_hooks.register_forward_hook(my_hook)
with global_hooks:
    # Call the modules here
```
I think this one is better. `torch.nn.Module` can check if the hooks are stored in the thread-local storage and then apply them without having some global state.

albanD commented:
Hi,

Thanks for all the answers!
Looks like it should be fairly easy to make it have the same behavior as regular hooks so that sounds good.

I think the global setter is fine if you set them but never remove them.
If you want to be able to nicely remove them, then the context manager looks better but might be more complex to implement and get right.

albanD commented on test/test_nn.py:
```diff
@@ -11392,6 +11390,175 @@ def __init__(self):
+            self.assertIn("bw_fail2", err.exception.args[0])
+            self.assertIn("got 2, but expected 1", err.exception.args[0])
+
+    def test_module_global_hook_writeable(self):
```
Can you mention in the test name that this one is for backward hook only and the one below for forward and pre_forward only. And use the same name: either writable or mutations. But since they test the same thing, keep the same name.

albanD commented on torch/nn/modules/module.py:
```diff
@@ -464,6 +470,43 @@ def convert(t):
+        computations. :attr:`grad_input` will only correspond to the inputs given
+        as positional arguments.
+
+        Global hooks is calles after hooks registered with `register_backward_hook`
```
nit: called

albanD commented on torch/nn/modules/module.py:
```diff
@@ -464,6 +470,43 @@ def convert(t):
 
+    @classmethod
+    def register_global_backward_hook(cls, hook):
+        r"""Registers a backward hook common to all the modules.
```
If this stays a global flag, I would add a Note or Warning here explaining that this sets a global state and the user should avoid using these if possible.
Recommended use cases is for debugging issues with nn.Module to have a better idea when they are called.

Same comment for the other two.

albanD commented on torch/nn/modules/module.py:
```diff
@@ -498,6 +541,33 @@ def register_backward_hook(self, hook):
+        single modified value in the hook. We will wrap the value into a tuple
+        if a single value is returned(unless that value is already a tuple).
+
+        This hook has precedence over the specific module hooks registered with
```
Why is this one executed before the local hooks while the two others are after?

emcastillo replied:
Actually this is one point I wanted to discuss, the ordering of execution.

Do you think that having the global before all the local ones is a reasonable behavior?
I will adjust it.

albanD commented on torch/nn/modules/module.py:
```diff
@@ -544,6 +641,17 @@ def register_forward_hook(self, hook):
         return handle
 
+    @classmethod
+    def clear_global_hooks(cls):
```
Why do we want this? The user already has the handles to clear the states if he wants to. This could potentially clear out other hooks that the user is not aware of.

emcastillo replied:
I use it mostly in the testing code to remove the global state between tests.
Should pretending `_` to it be ok? I can also recreate the dicts in the teardown code as well.

albanD commented on torch/nn/modules/module.py:
```diff
@@ -565,6 +673,12 @@ def _slow_forward(self, *input, **kwargs):
         return result
 
     def __call__(self, *input, **kwargs):
+        for hook in self._global_forward_pre_hooks.values():
```
This could be merged in the for loop below to re-use the code by chaining the two iterators.

Same for the two others below.

albanD reviewed (commented):
While I would see the benefit for a debug tool and for specific use, I am not sure that your example is especially convincing for me.

But doing something equivalent out of tree would be tricky so I think it is worth adding but I think the doc should mention that this should be used for debugging purposes primarily as setting global states in your code is not a good sign in general.
@mruberry what do you think?

mruberry commented:
> While I would see the benefit for a debug tool and for specific use, I am not sure that your example is especially convincing for me.
> 
> But doing something equivalent out of tree would be tricky so I think it is worth adding but I think the doc should mention that this should be used for debugging purposes primarily as setting global states in your code is not a good sign in general.
> @mruberry what do you think?

I really appreciate how thorough this PR's testing and documentation are, and I think @albanD has done a good cleanup review, too. I am also worried about setting and clearing global hooks, and agree that it's only appropriate for debugging and prototyping. We don't want, for example, this to become so common that library writers use it, because then there will be global conflicts between user hooks and library hooks, and clearing the global hooks will (as @albanD already mentions) have deleterious effects. 

Was this a common pattern when prototyping Chainer programs, @emcastillo?

emcastillo commented:
Thanks for the comments! Let me work on the review :)
I originally just replicated the same tests for the local hooks, so it might be good to revisit those aswell as I think that some of @albanD comments may apply there too. Should we do this in a different PR?

emcastillo commented:
Regarding the Chainer question. Yes, we provided global hooks to do profiling or debugging behavior as you can see in this case:
https://docs.chainer.org/en/stable/reference/generated/chainer.link_hooks.TimerHook.html#chainer.link_hooks.TimerHook

However, We used them as context managers instead of just storing a global list. 
```
from chainer.link_hooks import TimerHook
hook = TimerHook()
with hook:
    trainer.run()
hook.print_report()
```
I am fine with changing the design to obey this pattern, which I personally like more.

albanD commented:
How is the context manager handled in chainer? Using a TLS?
We don't have any TLS in python at the moment and it might be a bit tricky. Modules like DataParallel for example use multiple threads during the forward.

emcastillo commented:
Hi, Thanks for the follow-up.

Yes, in chainer we have a thread-local object

https://github.com/chainer/chainer/blob/df53bff3f36920dfea6b07a5482297d27b31e5b7/chainer/__init__.py#L129-L132

https://github.com/chainer/chainer/blob/v7.4.0/chainer/link_hook.py#L117-L119

I agree that if there is no current thread-local storage in PyTorch, it is not a good idea to add it just for this use case.

Should we proceed with the current design? If it's ok for you guys I will start working on the asked changes. Thanks!

albanD commented:
Yes I think we can proceed with the global state design.
Make sure in the doc to insist about the global state and the fact that this should be used for debugging only.

emcastillo commented on test/test_nn.py:
```diff
@@ -11392,6 +11390,175 @@ def __init__(self):
+
+    def test_module_global_hook_writeable(self):
+        self._clear_hooks()
+        module = nn.Linear(5, 5)
```
This is happening also in the regular hooks tests, we should fix these as well in a separate PR

emcastillo commented on test/test_nn.py:
```diff
@@ -11392,6 +11390,175 @@ def __init__(self):
+        self.assertEqual(counter['backwards'], 0)
+
+        output_1.backward(torch.ones(5, 5) * 2, retain_graph=True)
+        output_2.backward(torch.ones(5, 5) * 2, retain_graph=True)
```
exactly!

emcastillo commented on test/test_nn.py:
```diff
@@ -11392,6 +11390,175 @@ def __init__(self):
+        output = bn(torch.randn(5, 5, requires_grad=True))
+        output.sum().backward()
+
+    def test_module_global_hook_fail(self):
```
exactly, what is testing is that the tensors that the hooks return are correct or not.
This test is the same as the one for local hooks 😖

I am not sure whether this test is really needed, as it is only testing if the hook executed and changed the inputs to the functions or not.

Commit:
Fix tests

Commit:
Review changes and test of hooks ordering

emcastillo commented:
Hi, I just rebased and push the changes requested.
I added an extra test to verify the hooks are executed in the correct order.

I will apply this changes to the regular hooks tests too in a different PR since I just copied the tests from them :).

Thanks a lot for all the comments

albanD commented on torch/nn/modules/module.py:
```diff
@@ -494,6 +500,47 @@ def convert(t):
+        computations. :attr:`grad_input` will only correspond to the inputs given
+        as positional arguments.
+
+        Global hooks is called after hooks registered with `register_backward_hook`
```
Should this be updated to be before now?

albanD commented on torch/nn/modules/module.py:
```diff
@@ -553,6 +632,38 @@ def register_forward_pre_hook(self, hook: Callable[..., None]) -> RemovableHandl
+                a handle that can be used to remove the added hook by calling
+                ``handle.remove()``
+
+        This hook will be executed after specific module hooks registered with
```
Same ordering change

albanD reviewed (commented):
Thanks for the updates, it looks good.
Just doc updates and a small question:

emcastillo commented on test/test_nn.py:
```diff
@@ -11445,7 +11443,215 @@ def __init__(self):
+        def bw_fail2(self, grad_input, grad_output):
+            return grad_input + (torch.randn(2, 2),)
+
+        with nn.Sigmoid.register_global_backward_hook(bw_fail1):
```
Maybe something like `torch.nn.register_module_hook` is cleaner.

emcastillo replied:
I changed the hooks not to live in the nn.Module class but in the `nn.modules.module` module

Commit:
Change API and fix docs

albanD reviewed (approved):
Thanks for the update.
Looks good for me !

emcastillo commented:
Thanks!
